### ISSUE TITLE
[BUG] antivirus "message" is ignored

### ISSUE THREAD

**Issue by clangguth** [rspamd/rspamd#3279]
### Prerequisites

* [ X] Put an X between the brackets on this line if you have done all of the following:
    * Read about bug reporting in general: https://rspamd.com/doc/faq.html#how-to-report-bugs-found-in-rspamd
    * Enabled relevant debugging logs: https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd 
    * Checked the FAQs about Core files in case of fatal crash: https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed
    * Tried ASAN package and obtained the ASAN report (if possible): https://rspamd.com/doc/faq.html#asan-builds 
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+user%3Arspamd
    * Checked that there is not already an experimental package or master branch

**Describe the bug**

The `message` setting in `antivirus.conf` is ignored. Instead, a generic "Spam message rejected" message is returned to the client.

### Steps to Reproduce
```
# local.d/antivirus.conf snippet
clamav {
  [...]
  action = "reject";
  message = 'Virus found: ${VIRUS}';
  [...]
}
```

**Expected behavior**
milter-reject message "Virus found: Win.Test.EICAR_HDB-1"

**Actual behavior**
milter-reject message "Spam message rejected"

### Versions
Affected: Version 2.4 (precisely: `2.4-2~bionic` from `http://rspamd.com/apt-stable bionic/main`)

Note: exactly this configuration worked correctly (i.e., as expected) in version 2.2.

### Additional Information

This seems to be related, or at least very similar, to #3269 and #3276. However, from what I understand, these were presumably fixed in (or before) version 2.4, whereas this bug occurs with version 2.4-2.

**Comment by vstakhov** [comment]
You have not fulfilled the prerequisites for filling issues.

**Comment by clangguth** [comment]
Very helpful, thanks.

**Comment by bauerstefan** [comment]
Can confirm. Individual error messages are not honored anymore.
2.3-1~buster broke it on Debian 10.

rspamd logs correct error message, however the milter return to MTA, is an other:

`2020-03-03 21:17:01 #11692(normal) <6a3ba2>; task; rspamd_task_write_log: id: <!&!AAAAAAAAAAAuAAAAAAAAAHyq4G6e6XtOsHuLNaXnnBcBAMO2jhD3dRHOtM0AqgC7tuYAAAAAAA4AABAAAAAUEYqtiFbbQb854R0ciJXFAQAAAAA=@senderdomain.de>, qid: <0453E5DC93>, ip: 185.15.192.33, from: <mail@senderdomain.de>, (default: T (reject): [-5.72/500.00] [NEURAL_HAM_LONG(-3.00){-1.003;},BAYES_HAM(-3.00){100.00%;},NEURAL_HAM_MED(-1.07){-1.079;},AUTH_NA(1.00){},MANY_INVISIBLE_PARTS(1.00){10;},NEURAL_HAM_SHORT(-0.63){-1.064;},MIME_BAD_EXTENSION(0.20){pdf;},MIME_GOOD(-0.10){multipart/mixed;multipart/related;multipart/alternative;text/plain;},RCVD_IN_DNSWL_LOW(-0.10){185.15.192.33:from;},ARC_NA(0.00){},ASN(0.00){asn:15817, ipnet:185.15.192.0/22, country:DE;},DMARC_NA(0.00){senderdomain.de;},FILENAME_BLACKLISTED(0.00){zip;},FORCE_ACTION_FILENAME_BLACKLISTED(0.00){reject;},FROM_EQ_ENVFROM(0.00){},FROM_NO_DN(0.00){},HAS_ATTACHMENT(0.00){},MID_RHS_MATCH_FROM(0.00){},MIME_TRACE(0.00){0:+;1:+;2:+;3:+;4:~;5:~;6:~;7:-;...;},PREVIOUSLY_DELIVERED(0.00){john.doe@mydomain.tld;},RCPT_COUNT_ONE(0.00){1;},RCVD_COUNT_THREE(0.00){3;},RCVD_TLS_LAST(0.00){},RCVD_VIA_SMTP_AUTH(0.00){},R_DKIM_NA(0.00){},R_SPF_NA(0.00){no SPF record;},SOPHOS_VIRUS_FAIL(0.00){failed to scan and retransmits exceed;},TO_DN_ALL(0.00){},TO_MATCH_ENVRCPT_ALL(0.00){}]), len: 7763839, time: 1170.010ms, dns req: 38, digest: <d3b6cedbd80e2d81f617b7f88dc54946>, rcpts: <john.doe@mydomain.tld>, mime_rcpts: <john.doe@mydomain.tld>, **forced: reject "Attachment blocked. Only PDF is allowed. (cwERR502)"; score=nan (set by f`**

Mar  3 21:09:07 mx3 postfix/cleanup[26522]: 992ED5DC93: milter-reject: END-OF-MESSAGE from mailout02.de[185.15.191.31]: **5.7.1 Spam detected. (cwERR503**); from=<mail@senderdomain.de> to=<john.doe@mydomain.tld> proto=ESMTP helo=<mailout02.de>

The error message that is returned comes from

root@mx3: grep ERR503 /etc/rspamd/local.d/worker-proxy.inc 
reject_message = "Spam detected. (cwERR503)";

**Comment by bauerstefan** [comment]
Can confirm with 2.4-2~buster.

MTA receives default worker-proxy.inc reject-message:

root@mx3:~# grep ERR503 /etc/rspamd/local.d/worker-proxy.inc 
reject_message = "Spam detected. (cwERR503)";

Mar  3 22:36:58 mx2 postfix/cleanup[26967]: 69ACD7F2FD: milter-reject: END-OF-MESSAGE from sender[IP]: ... Spam detected. (cwERR503); from=<sender> to=<recipient> proto=ESMTP helo=<sender-server>


However rspamd logs correct force reject message.

2020-03-03 22:36:58 #28693(normal) <603954>; task; rspamd_task_write_log: id: <kcEE.YJ5ncKd/R3uGN/6CN8X3Gw.gNuZ2aPx1QE@edmkop03.bla.sender.tld>, qid: <69ACD7F2FD>, ip: sender-ip, from: <me@sender.tld>, (default: T (reject): [8.00/500.00] [BAYES_HAM(-2.99){99.99%;},NEURAL_HAM_LONG(-2.99){-0.998;},DWL_DNSWL_MED(-2.00){sender.tld:dkim;},MIME_BAD_EXTENSION(2.00){exe;},MIME_BAD_ATTACHMENT(1.60){exe;},NEURAL_HAM_MED(-1.20){-1.201;},TO_EXCESS_QP(1.20){},NEURAL_HAM_SHORT(-0.87){-1.455;},DMARC_POLICY_ALLOW(-0.50){sender.tld;reject;},MV_CASE(0.50){},RCVD_DKIM_ARC_DNSWL_MED(-0.50){},RCVD_IN_DNSWL_MED(-0.20){217.6.110.189:from;},R_DKIM_ALLOW(-0.20){sender.tld:s=18030501;},R_SPF_ALLOW(-0.20){+mx;},MIME_GOOD(-0.10){multipart/mixed;multipart/alternative;text/plain;},MIME_UNKNOWN(0.10){application/x-ms-dos-executable;},ARC_NA(0.00){},ASN(0.00){asn:3320, ipnet:217.0.0.0/13, country:DE;},CLAM_VIRUS(0.00){SecuriteInfo.com.Trojan.Generic.dlwwh.UNOFFICIAL;},DKIM_TRACE(0.00){sender.tld:+;},FILENAME_BLACKLISTED(0.00){exe;},FROM_EQ_ENVFROM(0.00){},FROM_HAS_DN(0.00){},HAS_ATTACHMENT(0.00){},HAS_X_PRIO_THREE(0.00){3;},MIME_TRACE(0.00){0:+;1:+;2:+;3:~;4:-;4:~;},PREVIOUSLY_DELIVERED(0.00){me@recipient.tld;},RCPT_COUNT_ONE(0.00){1;},RCVD_COUNT_THREE(0.00){3;},RCVD_TLS_LAST(0.00){},SOPHOS_VIRUS_FAIL(0.00){SAVDI: Message oversized (REJ 4);},TO_DN_EQ_ADDR_ALL(0.00){},TO_MATCH_ENVRCPT_ALL(0.00){}]), len: 1511199, time: 3874.865ms, dns req: 49, digest: <e237c525bef2141693e356758c17600f>, rcpts: <me@recipient.tld>, mime_rcpts: <me@recipient.tld>, forced: reject "Virus detected. (cwERR501)"; score=nan (set by clamav)

as set in 

root@mx2:~# grep 501 /etc/rspamd/local.d/*
/etc/rspamd/local.d/antivirus.conf:  message = 'Virus detected. (cwERR501)';

**Comment by vstakhov** [comment]
Please check with the experimental packages (as suggested by issue template btw).

**Comment by bauerstefan** [comment]
Thank you.. Need to build test-environment for that. Might take a few days. Will report back.

**Comment by bauerstefan** [comment]
sorry, can not test at the moment with experimental packages. @clangguth - do you have the chance to test?

**Comment by clangguth** [comment]
@bauerstefan: Sorry, no - I only have a production system with the official package.

Maybe I'm naïve, but I expect maintainers to take bug reports for officially released versions seriously, instead of simply dismissing them with snarky remarks, or requiring people to install experimental versions. Especially given that - IF the experimental version is already installed and properly testable, which I expect from any serious developer - it would take about 1 minute to verify if the bug is still present. Seems like I'm expecting too much though.

In the meantime, I'm using this workaround, which at least returns a meaningful message, albeit without the details of _which_ virus was found:
```
# /etc/rspamd/local.d/force_actions.conf

rules {
  ### WORKAROUND, see antivirus.conf
  CLAM_VIRUS_REJECT_WITH_MESSAGE {
    action = "reject";
    expression = "CLAM_VIRUS";
    message = "Malicious message rejected";
  }
}
```

**Comment by vstakhov** [comment]
I do not care about bug reports for released versions unless that are security issues. 

If you want to see your issues fixed then please use rolling releases that I recommend you to test.

**Comment by nafetsreuab** [comment]
@clangguth thank you for providing a workaround :+1: 
As rspamd is opensource, we can not expect or demand anyhting :) Given that rspamd looks like a one-man show, it's stable and we have regular updates.
@vstakhov thank you for your work and please keep up with it.

**Comment by chkpnt** [comment]
Just found the corresponding commit: https://github.com/rspamd/rspamd/commit/a345b74e1b32981f0a5bc1db7518c9ecc6c7b386

The issue is that the scanner results are handled as "least". So if the received mail isn't a virus only, but it's score also high enough to trigger a spam action, then the corresponding message will be shown.

If I revert the change in my local common.lua, a virus mail is rejected as expected:
```
          Subject: Undelivered Mail Returned to Sender
          ...
          I'm sorry to have to inform you that your message could not
          be delivered to one or more recipients. It's attached below.
          ...
          <jane.doe@mydomain.test>: host sut.mydomain.test[10.0.3.10] said: 554 5.7.1
              clamav: virus found: "Win.Test.EICAR_HDB-1" (in reply to end of DATA
              command)
          ...
```

But as a345b74e1b32981f0a5bc1db7518c9ecc6c7b386 was introduced to fix #3184 the flag should not simply be removed but be dependent on the action. I'll try to file a PR soon :-)

**Comment by chkpnt** [comment]
While PR #3342 fixes the bug described here, I honestly haven't fully understood how that pre-result is supposed to work. I played around and simulated different cases:

- if antivirus action is "add header", but the mail itself would had a score lower then the "add header" threshold defined in actions.conf, the score gets raised to that threshold and an "X-Spam" flag is added. I thought something like "X-Virus" would be added.
- if antivirus action is "rewrite subject", the subject gets rewritten. But not with the subject I define in antivirus.conf (`"***VIRUS*** %s"`), but with `"***SPAM*** %s"`.
